src/routes/maven.rs

```diff
@@ -0,0 +1,262 @@
+use crate::auth::get_user_from_headers;
+use crate::database;
+use crate::models::mods::ModId;
+use crate::routes::ApiError;
+use actix_web::{get, web, HttpRequest, HttpResponse};
+use sqlx::PgPool;
+use yaserde_derive::YaSerialize;
+
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(root = "metadata", rename = "metadata")]
+pub struct Metadata {
+    #[yaserde(rename = "groupId")]
+    group_id: String,
+    #[yaserde(rename = "artifactId")]
+    artifact_id: String,
+    versioning: Versioning,
+}
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(rename = "versioning")]
+pub struct Versioning {
+    latest: String,
+    release: String,
+    versions: Versions,
+    #[yaserde(rename = "lastUpdated")]
+    last_updated: String,
+}
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(rename = "versions")]
+pub struct Versions {
+    #[yaserde(rename = "version")]
+    versions: Vec<String>,
+}
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(rename = "project", namespace = "http://maven.apache.org/POM/4.0.0")]
+pub struct MavenPom {
+    #[yaserde(rename = "xsi:schemaLocation", attribute)]
+    schema_location: String,
+    #[yaserde(rename = "xmlns:xsi", attribute)]
+    xsi: String,
+    #[yaserde(rename = "modelVersion")]
+    model_version: String,
+    #[yaserde(rename = "groupId")]
+    group_id: String,
+    #[yaserde(rename = "artifactId")]
+    artifact_id: String,
+    version: String,
+    name: String,
+    description: String,
+}
+
+#[get("maven/modrinth/{id}/maven-metadata.xml")]
+pub async fn maven_metadata(
+    req: HttpRequest,
+    info: web::Path<(String,)>,
+    pool: web::Data<PgPool>,
+) -> Result<HttpResponse, ApiError> {
+    let string = info.into_inner().0;
+    let id_option: Option<ModId> = serde_json::from_str(&*format!("\"{}\"", string)).ok();
+
+    let mut mod_data;
+
+    if let Some(id) = id_option {
+        mod_data = database::models::Mod::get_full(id.into(), &**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
```

Comment:
If I recall correctly, we are using the `thiserror` crate, that allows us to skip entirely this mapping process, so you should be able to use the `?` shorthand directly after the await.

Reply:
I can't just use ? here because the `sqlx::error::Error` is turned into a `labrinth::database::models::DatabaseError` using the `?` operator, and then is wrapped by an `ApiError`. The `?` operator can't be used alone here because there isn't any direct conversion from `sqlx::error::Error` to `ApiError`.

Reply:
In that case, even if it's not the best looking thing, you could implement From<sqlx::error::Error> for ApiError that wraps everything in place.
It should be a 5 line thing, and that will make that part of the code far more readable.